examples/ASP.NET Core 2/Visual Studio 2017/ASP.NET Core 2 - VS2017/Program.cs

```diff
@@ -0,0 +1,41 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using NLog;
+using NLog.Config;
+using NLog.Web;
+
+namespace ASP.NET_Core_2___VS2017
+{
+    public class Program
+    {
+        public static void Main(string[] args)
+        {
+            // NLog: setup the logger first to catch all errors
+            var logger = NLogStart.InitLogger("NLog.config");
```

Comment:
@snakefoot inspiration for a better names?

Reply:
InitLogger is very confusing. It almost seem like you are creating a logger named "NLog.config".

Maybe rename NLogStart to NLogBuilder, and rename "InitLogger" to "LoadConfig" where LoadConfig returns NLogBuilder (fluent interface like)

NLogBuilder.LoadConfig("NLog.config").GetCurrentClassLogger()

NLogBuilder.GetCurrentClassLogger()

Reply:
Be aware that calling GetCurrentClassLogger within NLogBuilder will capture the wrong classname.

Reply:
> NLogBuilder.LoadConfig("NLog.config").GetCurrentClassLogger()

`LoadConfig` could return the `LogFactory`, not sure it that's clear. What do you think?

Reply:
> not sure it that's clear. What do you think?

Actually what you want is just the LogManager-interface, but you want people to perform the registration of AspNetExtensions before loading the config (From file or programmatically). Think it would be easier for all NLog-users if LogManager remained the default goto-place.

Think you want something like ConfigureNLog that just configures everything, and then people can use LogManager from there.

Maybe you should just change the RegisterExtendedItems() to attempt loading known good NLog-assemblies. Like NLog.Extended.dll and NLog.Web.dll and NLog.Web.AspNetCore.dll

Usually this auto-loading of NLog-dlls works out-of-the-box, but not when using Nuget-package-dlls that reside in different sub-folder.

Reply:
Maybe better that you just call it ConfigureNLog, but maybe you should change all ConfigureNLog-implementations to return the LogFactory object (instead of LoggingConfiguration). Since the LoggingConfiguration-object is kind of dull, and you can still access it through LogFactory.

Reply:
> Since the LoggingConfiguration-object is kind of dull, and you can still access it through LogFactory.

What's dull about it? That changing the config won't work until you reset the .config property?

Reply:
well Flush isn't needed in this case? Or do you think it's wise to add it after `BuildWebHost`?

Reply:
I can just see that people like the fluent-interface of Serilog, and it requires that you have access to a "powerful" object like the LogFactoy. But yes normally one will just be calling GetCurrentClassLogger().

But I prefer to have 3 stages. Startup - Running - Shutdown. Where Startup is where I load the NLog-config and the Running-stage where I acquire my loggers. Each stage happens at different locations in the code, so I don't need to build the nlog-config and get the logger-interface at the same location.

Reply:
We load it once here?

Fluent config would be great indeed. With the json config feature, that's one of the next steps for 4.6/4.7 I think

Reply:
> We load it once here?

Yes but one might think that if I need to get a logger at different place, then I should just do the same (Call LoadConfig/ConfigureNLog and then GetCurrentClassLogger), and then one might be surprised by the side-effects of the config-reload.

Or one might be cheated by the NLog-singleton-behavior, as it is hard to have multiple configs at the same time.